cmd/update-main.go

```diff
@@ -285,69 +279,75 @@ func DownloadReleaseData(timeout time.Duration) (data string, err *probe.Error)
 // fbe246edbd382902db9a4035df7dce8cb441357d mc.RELEASE.2016-10-07T01-16-39Z
 //
 // The second word must be `mc.` appended to a standard release tag.
-func parseReleaseData(data string) (sha256Hex string, releaseTime time.Time, err *probe.Error) {
+func parseReleaseData(data string) (sha256Hex string, releaseTime time.Time, releaseTag string, err *probe.Error) {
 	fields := strings.Fields(data)
 	if len(fields) != 2 {
-		return sha256Hex, releaseTime, probe.NewError(fmt.Errorf("Unknown release data `%s`", data))
+		return sha256Hex, releaseTime, "", probe.NewError(fmt.Errorf("Unknown release data `%s`", data))
 	}
 
 	sha256Hex = fields[0]
 	releaseInfo := fields[1]
 
 	fields = strings.SplitN(releaseInfo, ".", 2)
 	if len(fields) != 2 {
-		return sha256Hex, releaseTime, probe.NewError(fmt.Errorf("Unknown release information `%s`", releaseInfo))
+		return sha256Hex, releaseTime, "", probe.NewError(fmt.Errorf("Unknown release information `%s`", releaseInfo))
 	}
 	if fields[0] != "mc" {
-		return sha256Hex, releaseTime, probe.NewError(fmt.Errorf("Unknown release `%s`", releaseInfo))
+		return sha256Hex, releaseTime, "", probe.NewError(fmt.Errorf("Unknown release `%s`", releaseInfo))
 	}
 
 	releaseTime, err = releaseTagToReleaseTime(fields[1])
 	if err != nil {
-		return sha256Hex, releaseTime, err.Trace(fields...)
+		return sha256Hex, releaseTime, fields[1], err.Trace(fields...)
 	}
 
-	return sha256Hex, releaseTime, nil
+	return sha256Hex, releaseTime, fields[1], nil
 }
 
-func getLatestReleaseTime(timeout time.Duration) (sha256Hex string, releaseTime time.Time, err *probe.Error) {
-	data, err := DownloadReleaseData(timeout)
+func getLatestReleaseTime(customReleaseURL string, timeout time.Duration) (sha256Hex string, releaseTime time.Time, releaseTag string, err *probe.Error) {
+	data, err := DownloadReleaseData(customReleaseURL, timeout)
 	if err != nil {
-		return sha256Hex, releaseTime, err.Trace()
+		return sha256Hex, releaseTime, releaseTag, err.Trace()
 	}
 
 	return parseReleaseData(data)
 }
 
-func getDownloadURL(releaseTag string) (downloadURL string) {
+func getDownloadURL(customReleaseURL string, releaseTag string) (downloadURL string) {
 	// Check if we are docker environment, return docker update command
 	if IsDocker() {
 		// Construct release tag name.
 		return fmt.Sprintf("docker pull minio/mc:%s", releaseTag)
 	}
 
-	return mcReleaseURL + "archive/" + "mc." + releaseTag
+	u, err := url.Parse(customReleaseURL)
+	if err != nil {
```

Comment:
when we are supposed to get an error here ? shalln't we return it ?

Reply:
No it is on purpose url.Parse hardly fails with an error when it does it is safe to assume a fallback URL and attempt an `update` from main repo.